### PR TITLE
Reduce MergeMeshes resource utilization

### PR DESCRIPTION
**Pull request by ryantrem**
`MergeMeshes` tends to be heavily used in Babylon React Native to reduce the size of the scene graph since JS execution is relatively slow without JIT. However, `MergeMeshes` can be relatively slow (e.g. a few seconds when merging say 200-300 meshes for a glb, or a few hundred milliseconds when merging a few dozen meshes created at runtime, which can be a noticeable hitch). Given this, I spent a little time profiling `MergeMeshes` and improving several aspects of it:
- Rather than merge pairs of meshes repeatedly until all meshes are merged, just merge all meshes in one pass.
- Figure out how big we'll need the arrays to be up front and allocate them only once (the final size needed for all merged meshes).
- Don't use `Array.concat` as it is really slow and can't merge more than two arrays at once.

These changes make `MergeMeshes` about **5x-10x faster**, depending on the scenario. It also **reduces memory pressure about about 10x-12x**.

### Before
In this example, there were about 250 meshes. It took about 1.5 seconds and memory spiked up to about 1.4GB.
image

### After
Again, 250 meshes, but now taking about 233ms and memory climbing to only about 129mb (without big memory spikes and big GCs).
image

Unfortunately the change in perf was not as significant in Babylon Native, so I will have another round of changes coming that specifically targets native.

### CODE REVIEW

**Review by deltakosh — APPROVED** [review]

**Comment by sebavan** [comment]
Sounds like an amazing improvement, I ll take a closer look before merging !!!

File: src/Meshes/mesh.vertexData.ts [hunk 1]
```diff
@@ -464,103 +464,121 @@ export class VertexData {
...
-        if (!other) {
+        if (nonNullOthers.length === 0) {
             return source;
         }
 
-        var len = other.length + source.length;
-        var isSrcTypedArray = source instanceof Float32Array;
-        var isOthTypedArray = other instanceof Float32Array;
+        if (!source) {
+            return this._mergeElement(nonNullOthers[1], nonNullOthers.slice(1));
```
Comment (Popov72) [inline src/Meshes/mesh.vertexData.ts:555]: Shouldn't it be:
```typescript
            return this._mergeElement(nonNullOthers[0], nonNullOthers.slice(1));
```
?

File: src/Meshes/mesh.vertexData.ts [hunk 2]
```diff
@@ -464,103 +464,121 @@ export class VertexData {
...
+        }
 
-        // use non-loop method when the source is Float32Array
-        if (isSrcTypedArray) {
-            var ret32 = new Float32Array(len);
+        const len = source.length + nonNullOthers.reduce((othersSumLen, other) => othersSumLen += other.length, 0);
```
Comment (Popov72) [inline src/Meshes/mesh.vertexData.ts:558]: nit: same thing than above:
```typescript
nonNullOthers.reduce((curSum, other) => curSum + other.length, 0);
```

File: src/Meshes/mesh.ts [hunk 3]
```diff
@@ -4871,7 +4863,15 @@ export class Mesh extends AbstractMesh implements IGetSetVerticesData {
             }
         }
 
-        source = <Mesh>source;
+        const source = meshes[0];
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply src/Meshes/mesh.ts:4865]: Yes, you're right. I changed it to filter out null/undefined entries immediately and then removed the null checks. I think this should result in the same behavior as before.

File: src/Meshes/mesh.vertexData.ts (hunk 1 above)
Comment (ryantrem) [reply src/Meshes/mesh.vertexData.ts]: Wow how did I do that!? Glad you caught this one!

File: src/Meshes/mesh.ts [hunk 4]
```diff
@@ -4871,7 +4863,15 @@ export class Mesh extends AbstractMesh implements IGetSetVerticesData {
...
-        source = <Mesh>source;
+        const source = meshes[0];
+
+        const getVertexDataFromMesh = (mesh: Mesh) => {
+            const wm = mesh.computeWorldMatrix(true);
+            const vertexData = VertexData.ExtractFromMesh(mesh, true, true);
+            vertexData.transform(wm);
+            return vertexData;
+        };
+        const vertexData = getVertexDataFromMesh(source).merge(meshes.slice(1).map(mesh => getVertexDataFromMesh(mesh)));
```
[the remark this answers is not in the document]
Comment (ryantrem) [reply src/Meshes/mesh.ts]: Filtered out null/undefined meshes.

**Review by sebavan — APPROVED** [review]
I ll merge as soon as the build is green THX

File: src/Meshes/mesh.vertexData.ts (hunk 1 above)
Comment (sebavan) [reply src/Meshes/mesh.vertexData.ts]: @Popov72 King of reviews :-)

**Pull request merged**